src/SlackFunction.ts

```diff
@@ -46,83 +113,412 @@ export class SlackFunction {
+    });
+  };
+
+  /**
+   * Ensure that SlackFunction `complete()` utility callback is provided
+   * as args to all function and interactivity handlers registered on
+   * SlackFunction instance
+   * @param args middleware arguments
+   */
+  private prepareFnArgs(args: AnyMiddlewareArgs & AllMiddlewareArgs): AllSlackFunctionExecutedMiddlewareArgs {
+    const { next: _next, ...subArgs } = args;
+    // eslint-disable-next-line
+    const preparedArgs: any = { ...subArgs };
+    // ensure all handlers have complete utility
+    preparedArgs.complete = preparedArgs.complete ?? this.createComplete(preparedArgs);
+    return preparedArgs;
+  }
+
+  /**
+   * Creates a `complete()` utility function
+   *
+   * @param args
+   * @returns A `complete()` utility callback
+   * which can be accessed from any SlackFunction
+   * handler and used to to complete your SlackFunction.
+   *
+   * ```
+   * Example:
+   * const handler = async ({ complete }) => { complete() };
+   * const myFunc = new SlackFunction("id", handler);
+   * ```
+   */
+  // eslint-disable-next-line
+  private createComplete(args: any): CompleteFunction {
+    const { payload, body, client } = args;
+
+    // gets the function execution id from a function executed event
+    let { function_execution_id } = payload;
+
+    // gets the function execution id from the function data in a function interactivity event
+    if (function_execution_id === undefined && body !== undefined && 'function_data' in body) {
+      function_execution_id = body.function_data.execution_id;
     }
+
+    // if stil undefined, error
+    if (function_execution_id === undefined) {
+      const msg = '⚠️ Cannot generate required complete utility without a function_execution_id';
+      throw new SlackFunctionCompleteError(msg);
+    }
+
+    // return the utility callback
+    return ({ outputs, error }: CompleteFunctionArgs = {}) => {
+      // Slack API requires functions complete with either outputs or error, not both
+      if (outputs && error) {
+        throw new SlackFunctionCompleteError('⚠️ Cannot complete with outputs and error message supplied');
+      }
+      // if user has supplied outputs OR has supplied neither outputs nor error
+      if (outputs !== undefined || (outputs === undefined && error === undefined)) {
+        // helpful logging
+        this.logger.debug('🚀 Attempting to complete with outputs:', outputs);
+        return client.apiCall('functions.completeSuccess', {
+          outputs: outputs ?? {},
+          function_execution_id,
+        });
+      } if (error !== undefined) {
+        this.logger.debug('🚀 Attempting to complete with error:', error);
+        return client.apiCall('functions.completeError', {
+          error,
+          function_execution_id,
+        });
+      }
+      return null;
+    };
+  }
+}
+
+/* Event handling validation */
+
+export function isFunctionExecutedEvent(args: AnyMiddlewareArgs): boolean {
+  if (args.payload === undefined) {
```

Comment:
We may want to check for more than undefined (or maybe typescript guards us from this being an issue - if so, please ignore!), e.g. in case `payload` is `null`, then the next line would explode.